components/builder-db/src/models/channel.rs

```diff
@@ -264,18 +287,47 @@ pub struct OriginChannelDemote {
 impl OriginChannelPackage {
     pub fn promote(package: OriginChannelPromote, conn: &PgConnection) -> QueryResult<usize> {
         Counter::DBCall.increment();
-        diesel::sql_query("select * from promote_origin_package_v3($1, $2, $3)")
-            .bind::<Text, _>(package.origin)
-            .bind::<Text, _>(package.ident.to_string())
-            .bind::<Text, _>(package.channel)
-            .execute(conn)
+        // If this looks bad, it is. To ensure we get values here or die we have to execute queries to get the IDs first.
+        // I can hear the groaning already, "Why can't we just do a sub-select and let the database barf on insert"
+        // Great question - Because the typechecking happens in Rust and you wanted a type-safe language
```

Comment:
As discussed in the diesel gitter, this catches the possibility that one of the subqueries return no value. If that's impossible because of certain database constrains it is possible to trick diesel into accepting that query:

To do this you need to extend diesel by some custom type magic:

```rust

struct AssertNotNull<T, ST>(T, ::std::marker::PhantomData<ST>);

fn unsafe_assert_not_null<T, ST>(t: T) -> AssertNotNull<T, ST>
where T: Expression<SqlType = Nullable<ST>>,
      ST: NotNull
{
    AssertNotNull(t, Default::default())
}

impl<T, ST> Expression for AssertNotNull<T, ST>
where T: Expression<SqlType = Nullable<ST>>,
      ST: NotNull,
{
    type SqlType = ST;
}

impl<DB, T, ST> QueryFragment<DB> for AssertNotNull<T, ST>
where T: QueryFragment<DB>,
      DB: Backend
{

    fn walk_ast(&self, pass: AstPass<DB>) -> Result<(), Error> {
        self.0.walk_ast(pass)
    }
}
```

This defines some type that changes the sql type of a nullable expression to not nullable. After that just call `unsafe_assert_not_null` on both queries and it will work.